File: backend/app/routers/git.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional
import os
from pathlib import Path
from ..services.git_service import GitService, GitStatus, GitCommit, GitBranch, GitRemote

router = APIRouter(prefix="/git")

# Workspace directory
WORKSPACE_DIR = Path(os.getenv("WORKSPACE_DIR", "./workspace")).resolve()

class CommitRequest(BaseModel):
    message: str
    author: Optional[str] = None

class BranchRequest(BaseModel):
    name: str

class RemoteRequest(BaseModel):
    name: str
    url: str

class StageRequest(BaseModel):
    filepath: str

@router.get("/status", response_model=GitStatus)
async def get_status():
    """Get git repository status"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        return git_service.get_status()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/init")
async def init_repo():
    """Initialize a new git repository"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.init_repo():
            return {"message": "Repository initialized successfully"}
        else:
            raise HTTPException(status_code=500, detail="Failed to initialize repository")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/stage", response_model=dict)
async def stage_file(request: StageRequest):
    """Stage a file"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.stage_file(request.filepath):
            return {"message": "File staged successfully"}
        else:
            raise HTTPException(status_code=500, detail="Failed to stage file")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/unstage", response_model=dict)
async def unstage_file(request: StageRequest):
    """Unstage a file"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.unstage_file(request.filepath):
            return {"message": "File unstaged successfully"}
        else:
            raise HTTPException(status_code=500, detail="Failed to unstage file")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/commit", response_model=dict)
async def commit_changes(request: CommitRequest):
    """Commit staged changes"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.commit(request.message, request.author):
            return {"message": "Changes committed successfully"}
        else:
            raise HTTPException(status_code=500, detail="Failed to commit changes")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/commits", response_model=List[GitCommit])
async def get_commits(limit: int = 10):
    """Get recent commits"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        return git_service.get_commits(limit)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/branches", response_model=List[GitBranch])
async def get_branches():
    """Get all branches"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        return git_service.get_branches()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/branches", response_model=dict)
async def create_branch(request: BranchRequest):
    """Create a new branch"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.create_branch(request.name):
            return {"message": f"Branch '{request.name}' created successfully"}
        else:
            raise HTTPException(status_code=500, detail="Failed to create branch")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/branches/switch", response_model=dict)
async def switch_branch(request: BranchRequest):
    """Switch to a branch"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.switch_branch(request.name):
            return {"message": f"Switched to branch '{request.name}'"}
        else:
            raise HTTPException(status_code=500, detail="Failed to switch branch")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/remotes", response_model=List[GitRemote])
async def get_remotes():
    """Get remote repositories"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        return git_service.get_remotes()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/remotes", response_model=dict)
async def add_remote(request: RemoteRequest):
    """Add a remote repository"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.add_remote(request.name, request.url):
            return {"message": f"Remote '{request.name}' added successfully"}
        else:
            raise HTTPException(status_code=500, detail="Failed to add remote")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/pull", response_model=dict)
async def pull_changes(remote: str = "origin", branch: str = ""):
    """Pull from remote repository"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.pull(remote, branch):
            return {"message": "Pulled changes successfully"}
        else:
            raise HTTPException(status_code=500, detail="Failed to pull changes")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/push", response_model=dict)
async def push_changes(remote: str = "origin", branch: str = ""):
    """Push to remote repository"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if git_service.push(remote, branch):
            return {"message": "Pushed changes successfully"}
        else:
            raise HTTPException(status_code=500, detail="Failed to push changes")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/diff/{filepath:path}")
async def get_diff(filepath: str, staged: bool = False):
    """Get diff for a file"""
    try:
        git_service = GitService(WORKSPACE_DIR)
        if staged:
            diff = git_service.get_staged_diff(filepath)
        else:
            diff = git_service.get_diff(filepath)
        return {"diff": diff}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/api/git.ts
import axios from 'axios';

const API_BASE_URL = '/api/git';

export interface GitStatus {
  branch: string;
  clean: boolean;
  status: string[];
  ahead: number;
  behind: number;
}

export interface GitCommit {
  hash: string;
  message: string;
  author: string;
  date: string;
  short_hash: string;
}

export interface GitBranch {
  name: string;
  current: boolean;
  remote?: string;
}

export interface GitRemote {
  name: string;
  url: string;
}

// Create axios instance
const api = axios.create({
  baseURL: API_BASE_URL,
  timeout: 30000,
});

// Get repository status
export const getGitStatus = async (): Promise<GitStatus> => {
  try {
    const response = await api.get('/status');
    return response.data;
  } catch (error) {
    console.error('Error getting git status:', error);
    throw error;
  }
};

// Initialize repository
export const initRepo = async (): Promise<any> => {
  try {
    const response = await api.post('/init');
    return response.data;
  } catch (error) {
    console.error('Error initializing repository:', error);
    throw error;
  }
};

// Stage a file
export const stageFile = async (filepath: string): Promise<any> => {
  try {
    const response = await api.post('/stage', { filepath });
    return response.data;
  } catch (error) {
    console.error(`Error staging file ${filepath}:`, error);
    throw error;
  }
};

// Unstage a file
export const unstageFile = async (filepath: string): Promise<any> => {
  try {
    const response = await api.post('/unstage', { filepath });
    return response.data;
  } catch (error) {
    console.error(`Error unstaging file ${filepath}:`, error);
    throw error;
  }
};

// Commit changes
export const commitChanges = async (message: string, author?: string): Promise<any> => {
  try {
    const response = await api.post('/commit', { message, author });
    return response.data;
  } catch (error) {
    console.error('Error committing changes:', error);
    throw error;
  }
};

// Get recent commits
export const getCommits = async (limit: number = 10): Promise<GitCommit[]> => {
  try {
    const response = await api.get('/commits', { params: { limit } });
    return response.data;
  } catch (error) {
    console.error('Error getting commits:', error);
    throw error;
  }
};

// Get branches
export const getBranches = async (): Promise<GitBranch[]> => {
  try {
    const response = await api.get('/branches');
    return response.data;
  } catch (error) {
    console.error('Error getting branches:', error);
    throw error;
  }
};

// Create a new branch
export const createBranch = async (name: string): Promise<any> => {
  try {
    const response = await api.post('/branches', { name });
    return response.data;
  } catch (error) {
    console.error(`Error creating branch ${name}:`, error);
    throw error;
  }
};

// Switch to a branch
export const switchBranch = async (name: string): Promise<any> => {
  try {
    const response = await api.post('/branches/switch', { name });
    return response.data;
  } catch (error) {
    console.error(`Error switching to branch ${name}:`, error);
    throw error;
  }
};

// Get remotes
export const getRemotes = async (): Promise<GitRemote[]> => {
  try {
    const response = await api.get('/remotes');
    return response.data;
  } catch (error) {
    console.error('Error getting remotes:', error);
    throw error;
  }
};

// Add a remote
export const addRemote = async (name: string, url: string): Promise<any> => {
  try {
    const response = await api.post('/remotes', { name, url });
    return response.data;
  } catch (error) {
    console.error(`Error adding remote ${name}:`, error);
    throw error;
  }
};

// Pull changes
export const pullChanges = async (remote: string = "origin", branch: string = ""): Promise<any> => {
  try {
    const response = await api.post('/pull', null, { params: { remote, branch } });
    return response.data;
  } catch (error) {
    console.error('Error pulling changes:', error);
    throw error;
  }
};

// Push changes
export const pushChanges = async (remote: string = "origin", branch: string = ""): Promise<any> => {
  try {
    const response = await api.post('/push', null, { params: { remote, branch } });
    return response.data;
  } catch (error) {
    console.error('Error pushing changes:', error);
    throw error;
  }
};

// Get diff for a file
export const getDiff = async (filepath: string, staged: boolean = false): Promise<string> => {
  try {
    const response = await api.get(`/diff/${filepath}`, { params: { staged } });
    return response.data.diff;
  } catch (error) {
    console.error(`Error getting diff for ${filepath}:`, error);
    throw error;
  }
};
